registry/api/v2/urls.go

```diff
@@ -42,32 +42,46 @@ func NewURLBuilderFromString(root string) (*URLBuilder, error) {
 }
 
 // NewURLBuilderFromRequest uses information from an *http.Request to
-// construct the root url.
-func NewURLBuilderFromRequest(r *http.Request) *URLBuilder {
-	var scheme string
-
-	forwardedProto := r.Header.Get("X-Forwarded-Proto")
-
-	switch {
-	case len(forwardedProto) > 0:
-		scheme = forwardedProto
-	case r.TLS != nil:
-		scheme = "https"
-	case len(r.URL.Scheme) > 0:
-		scheme = r.URL.Scheme
-	default:
-		scheme = "http"
-	}
+// construct the root url. configScheme and configHost are the URI scheme and
+// hostname, respectively, from the configuration - or empty strings if
+// unspecified.
+func NewURLBuilderFromRequest(r *http.Request, configScheme, configHost string) *URLBuilder {
```

Comment:
Rather than crushing these together, why not have a `NewURLBuilderFromBase` method? The use case is completely disjoint.